libs/use-viewer/src/index.ts

```diff
@@ -89,127 +39,186 @@ const useViewer = ({ streamName, streamAccountId, subscriberToken, handleError }
-          if (!source.sourceId) source.sourceId = missedSourceId;
-          try {
-            const trackSource = await addRemoteTrackAndProject(source.sourceId, source.tracks, viewer.current);
-            setTimeout(() => {
-              dispatch({ type: ViewerActionType.ADD_SOURCE, source: trackSource });
-            }, 1000);
-          } catch (error) {
-            handleInternalError(error);
-          }
+        try {
+          const newRemoteTrackSource = await addRemoteTrackAndProject(viewer, sourceId, tracks);
+
+          setTimeout(() => {
+            dispatch({ remoteTrackSource: newRemoteTrackSource, sourceId, type: ViewerActionType.ADD_SOURCE });
+          }, 1000);
+        } catch (error) {
+          handleInternalError(error);
         }
         break;
-      case 'inactive':
-        {
-          const source = event.data as MediaStreamSource;
-          if (!source.sourceId) source.sourceId = missedSourceId;
-          dispatch({ type: ViewerActionType.REMOVE_SOURCE, sourceId: source.sourceId, viewer: viewer.current });
+
+      case 'inactive': {
+        const remoteTrackSource = remoteTrackSources.get(sourceId);
+
+        if (remoteTrackSource) {
+          unprojectRemoteTrackSource(viewer, remoteTrackSource);
+
+          dispatch({ type: ViewerActionType.REMOVE_SOURCE, sourceId });
         }
+
         break;
+      }
+
       case 'viewercount':
         setViewerCount((event.data as ViewerCount).viewercount);
         break;
+
       case 'layers': {
         dispatch({
           medias: (event.data as MediaStreamLayers).medias,
-          type: ViewerActionType.UPDATE_SOURCES_QUALITIES,
-          viewer: viewer.current as View,
+          type: ViewerActionType.UPDATE_SOURCES_QUALITY_OPTIONS,
         });
         break;
       }
     }
-  }, []);
+    return;
+  };
 
-  const connect = async () => {
-    if (!viewer.current) return;
-    console.log('connecting');
-    try {
-      await viewer.current.connect({ events: ['active', 'inactive', 'layers', 'viewercount'] });
-    } catch (error) {
-      console.error(error);
-      viewer.current?.reconnect();
-    } finally {
-      console.log('register broadcastEvent');
-      viewer.current?.webRTCPeer?.initStats();
-      viewer.current?.webRTCPeer?.on('stats', (statistics: StreamStats) => {
-        dispatch({
-          statistics,
-          type: ViewerActionType.UPDATE_SOURCES_STATISTICS,
-          viewer: viewer.current as View,
-        });
-      });
+  const handleStats = (statistics: StreamStats) => {
+    dispatch({
+      statistics,
+      type: ViewerActionType.UPDATE_SOURCES_STATISTICS,
+    });
+  };
+
+  // Get the audio/video media IDs for the main stream from the track event
+  // and use them to create the main stream mapping
+  const handleTrack = (event: RTCTrackEvent) => {
+    const {
+      streams,
+      track: { kind },
+      transceiver: { mid },
+    } = event;
+
+    if (streams.length && mid !== null) {
```

Comment:
The streams[0] here is the main stream that we must provide to App, this is the MediaStream which has 2MB bandwidth,  we are projecting other sources to this stream. This is critical!

Reply:
I guess you misunderstood, each source's MediaStream is different from the Main Stream which is stream[0] here. This probably is the root cause of why the MediaStream became inactive, you used the same MediaStream in two VideoViews but I am not 100% sure about this.

Reply:
Thanks for clarifying. This change has resolved the bug where switching between sources would cause their `mediaStream.active` property to go from `true` to `false.